src/library/Hooks/usePayouts/usePayouts.tsx

```diff
@@ -0,0 +1,163 @@
+/*
+ * Warning: lots of type assertions due to mistyped api library.
+ */
+import { useApi } from 'contexts/Api';
+import { useConnect } from 'contexts/Connect';
+import { useNetworkMetrics } from 'contexts/Network';
+import React, {
+  ReactElement,
+  useContext,
+  useEffect,
+  useMemo,
+  useReducer,
+  useState,
+} from 'react';
+import { round } from 'Utils';
+import asyncSequenceMap from './asyncSequenceMap';
+import calcPayoutForEra from './calcPayoutForEra';
+import fetchDataForEra from './fetchDataForEra';
+import { EraData } from './types';
+
+type ValidatorsData = {
+  [era: number]: EraData;
+};
+
+type ContextType = {
+  validatorsDataCache: ValidatorsData;
+  updateValidatorsDataCache: (patch: ValidatorsData) => void;
+};
+
+const Context = React.createContext<ContextType>({} as ContextType);
+
+/**
+ * Calculates payouts (aka rewards) based on stake data fetched from
+ * the chain. The data is aggressively cached, as it's historical and
+ * immutable. The returned values are in TAZERO (AZERO / 10^12).
+ *
+ * @param lastNEras Fetch last "lastNEras" eras, starting from the first
+ * finished one (the one behind the active one)
+ */
+export default (lastNEras: number) => {
+  const { validatorsDataCache, updateValidatorsDataCache } =
+    useContext(Context);
+  const { activeAccount: validatorId } = useConnect();
+  const { api } = useApi();
+
+  const {
+    metrics: {
+      activeEra: { index: activeEra },
+    },
+  } = useNetworkMetrics();
+  const { selectedErasDataFromCache, erasMissingInCache } = useMemo(
```

Comment:
This caching logic may seem as an exaggeration, but considering how slow it is to fetch all the data, I reckoned it's a necessity. With this logic, no era data is fetched twice no matter where you go on the page, until you leave it / reload.